src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/QueueItem.cs

```diff
@@ -0,0 +1,169 @@
+            mutatesSolutionState,
+            methodName,
+            methodHandler,
+            request,
+            handler,
+            lspServices,
+            logger,
+            cancellationToken);
+
+        return (queueItem, queueItem._completionSource.Task);
+    }
+
+    /// <summary>
+    /// Processes the queued request. Exceptions will be sent to the task completion source
+    /// representing the task that the client is waiting for, then re-thrown so that
+    /// the queue can correctly handle them depending on the type of request.
+    /// </summary>
+    /// <param name="cancellationToken"></param>
+    /// <returns>The result of the request.</returns>
+    public async Task StartRequestAsync(CancellationToken cancellationToken)
+    {
+        _logger.LogStartContext($"{MethodName}");
+        try
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var requestContextFactory = LspServices.GetRequiredService<IRequestContextFactory<TRequestContext>>();
+            var context = await requestContextFactory.CreateRequestContextAsync(this, _request, cancellationToken).ConfigureAwait(false);
+
+            if (context is null)
+            {
+                // If we weren't able to get a corresponding context for this request (for example, we
+                // couldn't map a doc request to a particular Document, or we couldn't find an appropriate
+                // Workspace for a global operation), then just immediately complete the request with a
+                // 'null' response.  Note: the lsp spec was checked to ensure that 'null' is valid for all
+                // the requests this could happen for.  However, this assumption may not hold in the future.
+                // If that turns out to be the case, we could defer to the individual handler to decide
+                // what to do.
+                _logger.LogWarning($"Could not get request context for {MethodName}");
+                _completionSource.TrySetException(new InvalidOperationException($"Unable to create request context for {MethodName}"));
+            }
+            else
+            {
+                if (_handler is IRequestHandler<TRequest, TResponse, TRequestContext> requestHandler)
+                {
+                    var result = await requestHandler.HandleRequestAsync(_request, context, cancellationToken).ConfigureAwait(false);
+
+                    _completionSource.TrySetResult(result);
+                }
+                else if (_handler is INotificationHandler<TRequest, TRequestContext> notificationHandler)
+                {
+                    await notificationHandler.HandleNotificationAsync(_request, context, cancellationToken).ConfigureAwait(false);
+
+                    // We know that the return type of <see cref="INotificationHandler{TRequestType, RequestContextType}"/> will always be <see cref="VoidReturn" /> even if the compiler doesn't.
```

Comment:
I went back and forth on this. I opted for "no" because I wanted to protect the area most users will interact with (the handlers of individual LSP's) from this weird little internal thing that even a Notification is returning an object of some kind, at some level. Ideally we wouldn't have a `VoidReturn` type at all, so I tried to avoid it leaking out of this impl.